### PR TITLE
gamedata: Implemented OpenGL SSR reflections on water from LVutner

### PR DESCRIPTION
**Pull request by eagleivg**
Please check for possible errors or places that can be optimized.

### CODE REVIEW

File: res/gamedata/shaders/gl/ssr.h [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+	float4 screen = proj;
+	screen.x = (proj.x + proj.w);
+	screen.y = (proj.w - proj.y);
+	screen.xy *= 0.5;
+	return screen;
+}
+
+half is_sky(float depth)
+{
+	return step(abs(depth - 10000.f), 0.001);
```
Comment (SkyLoaderr) [inline res/gamedata/shaders/gl/ssr.h]:
```suggestion
	return step(abs(depth - 10000.0), 0.001);
```

File: res/gamedata/shaders/gl/ssr.h [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+}
+
+float3 hash(float3 a)
+{
+	a *= timers.x;
+    a = frac(a * 0.8f);
```
Comment (SkyLoaderr) [inline res/gamedata/shaders/gl/ssr.h]:
```suggestion
    a = frac(a * 0.8);
```

File: res/gamedata/shaders/gl/ssr.h [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+    a += dot(a, a.yxz + 19.19);
+    return frac((a.xxy + a.yxx)*a.zyx);
+}
+
+/*World space components as input*/
+#ifndef SSR_QUALITY
+float4 compute_ssr(float3 position, float3 normal, float3 skybox)
+{
+	return float4(skybox.xyz, 1.0f);
```
Comment (SkyLoaderr) [inline res/gamedata/shaders/gl/ssr.h]:
```suggestion
	return float4(skybox.xyz, 1.0);
```

File: res/gamedata/shaders/gl/ssr.h [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+}
+
+#else
+
+float4 compute_ssr(float3 position, float3 normal, float3 skybox)
+{
+	/*Initialize step size and error*/
+	float step = 1.0f/float(SSR_SAMPLES);
```
Comment (SkyLoaderr) [inline res/gamedata/shaders/gl/ssr.h]:
```suggestion
	float step = 1.0/float(SSR_SAMPLES);
```

File: res/gamedata/shaders/gl/ssr.h [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+		/*Prepare new position*/
+		float3 new_position = position + vreflect * step;
+
+		/*Add hash to new position*/
+	#ifdef SSR_JITTER
+		new_position += hash(position.xyz) * SSR_JITTER_INTENSITY;
+	#endif
+
+		/*Convert new position to texcoord*/
+		float4 proj_position = mul(m_VP, float4(new_position, 1.f));
```
Comment (SkyLoaderr) [inline res/gamedata/shaders/gl/ssr.h]:
```suggestion
		float4 proj_position = mul(m_VP, float4(new_position, 1.0));
```

File: res/gamedata/shaders/gl/ssr.h [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+		#else
+            float2 refl_tc_t = refl_tc.xy * screen_res.xy;
+			float hit_depth = texelFetch(s_position, int2(refl_tc_t.x, screen_res.y - refl_tc_t.y), 0).z;
+		#endif
+	#else
+			float hit_depth = tex2D(s_half_depth, float2(refl_tc.x, -refl_tc.y)).x;
+	#endif
+
+		/*Intersect sky from hit depth*/
+		hit_depth = lerp(hit_depth, 0.f, is_sky(hit_depth));
```
Comment (SkyLoaderr) [inline res/gamedata/shaders/gl/ssr.h]:
```suggestion
		hit_depth = lerp(hit_depth, 0.0, is_sky(hit_depth));
```

File: res/gamedata/shaders/gl/ssr.h [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+
+		/*Sample depth*/
+		float depth = mul(m_V, float4(position, 1.f)).z;
```
Comment (SkyLoaderr) [inline res/gamedata/shaders/gl/ssr.h]:
```suggestion
		float depth = mul(m_V, float4(position, 1.0)).z;
```

**Comment by SkyLoaderr** [comment]
Please, replace all `.f` -> `.0`

File: src/Layers/xrRenderPC_GL/rgl_shaders.cpp [hunk 8]
```diff
@@ -395,6 +396,20 @@ HRESULT CRender::shader_compile(pcstr name, IReader* fs, pcstr pFunctionName,
         appendShaderOption(softWater, "USE_SOFT_WATER", "1");
     }
 
+    // Water reflections
+	if (RImplementation.o.advancedpp && ps_r_water_reflection)
```
Comment (Xottab-DUTY) [inline src/Layers/xrRenderPC_GL/rgl_shaders.cpp]:
```suggestion
    if (RImplementation.o.advancedpp && ps_r_water_reflection)
```

File: src/Layers/xrRenderPC_GL/rgl_shaders.cpp [hunk 9]
```diff
@@ -395,6 +396,20 @@ HRESULT CRender::shader_compile(pcstr name, IReader* fs, pcstr pFunctionName,
...
+        xr_sprintf(c_water_reflection, "%d", ps_r_water_reflection);
+        options.add("SSR_QUALITY", c_water_reflection);
+        sh_name.append(ps_r_water_reflection);
+        const bool sshHalfDepth = ps_r2_ls_flags_ext.test(R3FLAGEXT_SSR_HALF_DEPTH);
+        appendShaderOption(sshHalfDepth, "SSR_HALF_DEPTH", "1");
+        const bool ssrJitter = ps_r2_ls_flags_ext.test(R3FLAGEXT_SSR_JITTER);
+        appendShaderOption(ssrJitter, "SSR_JITTER", "1");
+    }
+    else
+        sh_name.append(static_cast<u32>(0));
```
Comment (Xottab-DUTY) [inline src/Layers/xrRenderPC_GL/rgl_shaders.cpp:412]:
```suggestion
{
        sh_name.append(static_cast<u32>(0));
}
```

File: src/Layers/xrRenderPC_GL/rgl_shaders.cpp [hunk 10]
```diff
@@ -395,6 +396,22 @@ HRESULT CRender::shader_compile(pcstr name, IReader* fs, pcstr pFunctionName,
...
+        sh_name.append(ps_r_water_reflection);
+        const bool sshHalfDepth = ps_r2_ls_flags_ext.test(R3FLAGEXT_SSR_HALF_DEPTH);
+        appendShaderOption(sshHalfDepth, "SSR_HALF_DEPTH", "1");
+        const bool ssrJitter = ps_r2_ls_flags_ext.test(R3FLAGEXT_SSR_JITTER);
+        appendShaderOption(ssrJitter, "SSR_JITTER", "1");
+    }
+    else
+{
+        sh_name.append(static_cast<u32>(0));
+}
```
Comment (Xottab-DUTY) [inline src/Layers/xrRenderPC_GL/rgl_shaders.cpp]:
```suggestion
    }
```

File: src/Layers/xrRenderPC_GL/rgl_shaders.cpp [hunk 11]
```diff
@@ -395,6 +396,22 @@ HRESULT CRender::shader_compile(pcstr name, IReader* fs, pcstr pFunctionName,
...
+        xr_sprintf(c_water_reflection, "%d", ps_r_water_reflection);
+        options.add("SSR_QUALITY", c_water_reflection);
+        sh_name.append(ps_r_water_reflection);
+        const bool sshHalfDepth = ps_r2_ls_flags_ext.test(R3FLAGEXT_SSR_HALF_DEPTH);
+        appendShaderOption(sshHalfDepth, "SSR_HALF_DEPTH", "1");
+        const bool ssrJitter = ps_r2_ls_flags_ext.test(R3FLAGEXT_SSR_JITTER);
+        appendShaderOption(ssrJitter, "SSR_JITTER", "1");
+    }
+    else
+{
```
Comment (Xottab-DUTY) [inline src/Layers/xrRenderPC_GL/rgl_shaders.cpp]:
```suggestion
    {
```

**Review by Xottab-DUTY — COMMENTED** [review]
Oops, sorry, I did the wrong formatting in the previous review.

File: src/Layers/xrRender/xrRender_console.cpp [hunk 12]
```diff
@@ -61,6 +61,10 @@ const xr_token qsun_quality_token[] = {{"st_opt_low", 0}, {"st_opt_medium", 1},
 #endif // !USE_DX9
     {nullptr, 0}};
 
+u32 ps_r_water_reflection = 3;
+const xr_token qwater_reflection_quality_token[] = {{"st_opt_off", 0}, {"st_opt_low", 1}, {"st_opt_medium", 2},
+    {"st_opt_high", 3}, {"st_opt_ultra", 4}, {nullptr, 0}};
```
Comment (Xottab-DUTY) [inline src/Layers/xrRender/xrRender_console.cpp]:
```suggestion
const xr_token qwater_reflection_quality_token[] =
{
    { "st_opt_off", 0 },
    { "st_opt_low", 1 },
    { "st_opt_medium", 2 },
    { "st_opt_high", 3 },
    { "st_opt_ultra", 4 },
    { nullptr, -1 }
};
```

File: res/gamedata/shaders/gl/iostructs/p_water.h [hunk 13]
```diff
@@ -15,21 +15,23 @@ struct v2p
 	float3	v2point	; // TEXCOORD6;
 #if defined(USE_SOFT_WATER) && defined(NEED_SOFT_WATER)
 	float4	tctexgen; // TEXCOORD7;
-#endif	// defined(USE_SOFT_WATER) && defined(NEED_SOFT_WATER)	
+#endif	// defined(USE_SOFT_WATER) && defined(NEED_SOFT_WATER)
+	float4	position_w;	// POSITION0;
```
Comment (Xottab-DUTY) [inline res/gamedata/shaders/gl/iostructs/p_water.h:20]:
```suggestion
#if SSR_QUALITY > 0
	float4	position_w;	// POSITION0;
#endif
```

File: res/gamedata/shaders/gl/iostructs/p_water.h [hunk 14]
```diff
@@ -15,21 +15,23 @@ struct v2p
...
-layout(location = TEXCOORD6) 		in float3	v2p_water_v2point	; // TEXCOORD6;	
+layout(location = TEXCOORD2) 		in float2	v2p_water_tnorm1	; // TEXCOORD2;
+layout(location = TEXCOORD3) 		in float3	v2p_water_M1		; // TEXCOORD3;
+layout(location = TEXCOORD4) 		in float3	v2p_water_M2		; // TEXCOORD4;
+layout(location = TEXCOORD5) 		in float3	v2p_water_M3		; // TEXCOORD5;
+layout(location = TEXCOORD6) 		in float3	v2p_water_v2point	; // TEXCOORD6;
 #if defined(USE_SOFT_WATER) && defined(NEED_SOFT_WATER)
 layout(location = TEXCOORD7) 		in float4	v2p_water_tctexgen	; // TEXCOORD7;
 #endif	// defined(USE_SOFT_WATER) && defined(NEED_SOFT_WATER)
+layout(location = POSITION0) 		in float4	v2p_water_pos	; // POSITION0;
```
Comment (Xottab-DUTY) [inline res/gamedata/shaders/gl/iostructs/p_water.h:37]:
```suggestion
#if SSR_QUALITY > 0
layout(location = POSITION0) 		in float4	v2p_water_pos	; // POSITION0;
#endif
```

File: res/gamedata/shaders/gl/iostructs/p_water.h [hunk 15]
```diff
@@ -53,6 +55,7 @@ void main()
 #if defined(USE_SOFT_WATER) && defined(NEED_SOFT_WATER)
 	I.tctexgen	= v2p_water_tctexgen;
 #endif	// defined(USE_SOFT_WATER) && defined(NEED_SOFT_WATER)
+    I.position_w= v2p_water_pos;
```
Comment (Xottab-DUTY) [inline res/gamedata/shaders/gl/iostructs/p_water.h]:
```suggestion
#if SSR_QUALITY > 0
    I.position_w = v2p_water_pos;
#endif
```

File: res/gamedata/shaders/gl/iostructs/v_water.h [hunk 16]
```diff
@@ -22,7 +22,8 @@ struct v2p
 	float3	v2point	;	// TEXCOORD6;
 #if defined(USE_SOFT_WATER) && defined(NEED_SOFT_WATER)
 	float4	tctexgen;	// TEXCOORD7;
-#endif	// defined(USE_SOFT_WATER) && defined(NEED_SOFT_WATER)	
+#endif	// defined(USE_SOFT_WATER) && defined(NEED_SOFT_WATER)
+	float4	position_w;	// POSITION0;
```
Comment (Xottab-DUTY) [inline res/gamedata/shaders/gl/iostructs/v_water.h:27]:
```suggestion
#if SSR_QUALITY > 0
	float4	position_w;	// POSITION0;
#endif
```

File: res/gamedata/shaders/gl/iostructs/v_water.h [hunk 17]
```diff
@@ -36,14 +37,15 @@ layout(location = TEXCOORD0)		in float2	v_vert_uv		; // TEXCOORD0;		// (u0,v0)
...
-layout(location = TEXCOORD6) 		out float3	v2p_vert_v2point	; // TEXCOORD6;	
+layout(location = TEXCOORD2) 		out float2	v2p_vert_tnorm1		; // TEXCOORD2;
+layout(location = TEXCOORD3) 		out float3	v2p_vert_M1		; // TEXCOORD3;
+layout(location = TEXCOORD4) 		out float3	v2p_vert_M2		; // TEXCOORD4;
+layout(location = TEXCOORD5) 		out float3	v2p_vert_M3		; // TEXCOORD5;
+layout(location = TEXCOORD6) 		out float3	v2p_vert_v2point	; // TEXCOORD6;
 #if defined(USE_SOFT_WATER) && defined(NEED_SOFT_WATER)
 layout(location = TEXCOORD7) 		out float4	v2p_vert_tctexgen	; // TEXCOORD7;
 #endif	// defined(USE_SOFT_WATER) && defined(NEED_SOFT_WATER)
+layout(location = POSITION0) 		out float4	v2p_vert_pos	; // POSITION0;
```
Comment (Xottab-DUTY) [inline res/gamedata/shaders/gl/iostructs/v_water.h:51]:
```suggestion
#if SSR_QUALITY > 0
layout(location = POSITION0) 		out float4	v2p_vert_pos	; // POSITION0;
#endif
```

File: res/gamedata/shaders/gl/iostructs/v_water.h [hunk 18]
```diff
@@ -73,5 +75,6 @@ void main()
 #endif	// defined(USE_SOFT_WATER) && defined(NEED_SOFT_WATER)
 	v2p_vert_c0	= O.c0;
 	v2p_vert_fog	= O.fog;
+	v2p_vert_pos =  O.position_w;
```
Comment (Xottab-DUTY) [inline res/gamedata/shaders/gl/iostructs/v_water.h]:
```suggestion
#if SSR_QUALITY > 0
	v2p_vert_pos	=  O.position_w;
#endif
```

File: res/gamedata/shaders/gl/water.vs [hunk 19]
```diff
@@ -19,6 +19,7 @@ v2p _main (v_vert v)
 		P	= watermove(P);
 
 	o.v2point	= P.xyz-eye_position;
+	o.position_w= P;
```
Comment (Xottab-DUTY) [inline res/gamedata/shaders/gl/water.vs:23]:
```suggestion
#if SSR_QUALITY > 0
	o.position_w= P;
#endif
```

**Review by Xottab-DUTY — COMMENTED** [review]
Don't forget to hide the inclusion of SSR.h in water.ps:
```cpp
#if SSR_QUALITY > 0
#include "ssr.h"
#endif
```

File: src/Layers/xrRender/xrRender_console.cpp (hunk 12 above)
Comment (eagleivg) [reply src/Layers/xrRender/xrRender_console.cpp]: Are you sure? It violates the current code style in the file and will look alien.

Comment (Xottab-DUTY) [reply src/Layers/xrRender/xrRender_console.cpp]: Yes, I'm sure.

**Pull request merged**